nixos/modules/services/video/epgstation.nix

```diff
@@ -0,0 +1,335 @@
+
+          <warning>
+            <para>
+              This stores the plaintext password in the world-readable Nix
+              store.  Consider using <option>database.passwordFile</option>
+              instead.
+            </para>
+          </warning>
+        '';
+      };
+
+      passwordFile = mkOption {
+        type = types.path;
+        default = pkgs.writeText "epgstation-db-password" cfg.database.password;
+        example = "/run/keys/epgstation-db-password";
+        description = ''
+          A file containing the password corresponding to the database
+          <option>database.name</option>.
+        '';
+      };
+
+      connectTimeout = mkOption {
+        type = types.ints.unsigned;
+        default = 1000;
+        description = "Connection timeout for the database in milliseconds.";
+      };
+
+      connectionLimit = mkOption {
+        type = types.ints.unsigned;
+        default = 10;
+        description = "Maximum number of connections to the database.";
+      };
+    };
+
+    encode = mkOption {
+      type = with types; listOf attrs;
+      description = "Encoding presets for recorded videos.";
+      default = [
+        { name = "H264";
+          cmd = "${pkgs.epgstation}/libexec/enc.sh main";
+          suffix = ".mp4";
+          default = true; }
+        { name = "H264-sub";
+          cmd = "${pkgs.epgstation}/libexec/enc.sh sub";
+          suffix = "-sub.mp4"; }
+      ];
+    };
+
+    settings = mkOption {
+      type = settingsFmt.type;
+      default = {};
+      example = {
+        recPriority = 20;
+        conflictPriority = 10;
+      };
+      description = ''
+        Options which are addded to config.json.
+
+        Documentation:
+        <link xlink:href="https://github.com/l3tnun/EPGStation/blob/master/doc/conf-manual.md"/>
+      '';
+    };
+
+  };
+
+  config = mkIf cfg.enable {
+    environment.etc = {
+      "epgstation/operatorLogConfig.json".text = builtins.toJSON logConfig;
+      "epgstation/serviceLogConfig.json".text = builtins.toJSON logConfig;
+    };
+
+    networking.firewall = mkIf cfg.openFirewall {
+      # port (serverPort + 1) is used for websockets
+      allowedTCPPorts = with cfg; [ serverPort (serverPort + 1) ];
+    };
+
+    users.users.epgstation = {
+      description = "EPGStation user";
+      group = config.users.groups.epgstation.name;
+      isSystemUser = true;
+    };
+
+    users.groups.epgstation = {};
+
+    services.mirakurun.enable = mkDefault true;
+
+    services.mysql = {
+      enable = mkDefault true;
+      package = mkDefault pkgs.mysql;
+      ensureDatabases = [ cfg.database.name ];
+      # FIXME: enable once mysqljs supports auth_socket
+      # ensureUsers = [ {
+      #   name = username;
+      #   ensurePermissions = { "${cfg.database.name}.*" = "ALL PRIVILEGES"; };
+      # } ];
+    };
+
+    services.epgstation.settings = let
+      defaultSettings = {
+        inherit (cfg) readOnlyOnce serverPort mirakurunPath encode;
```

Comment:
Consider using the recently-introduced freeform modules: #82743 
They allow you to change the `settings` type to expose such options directly, like
```nix
{
  settings = mkOption {
    type = types.submodule {
      freeformType = settingsFmt.type;
      options.readOnlyOnce = mkOption { ... };
      options.serverPort = mkOption { ... };
      # ...
    };
  };
}
```

Then there's only one way to set these options through `cfg.settings.readOnlyOnce` (and not through both `cfg.readOnlyOnce` and `cfg.settings.readOnlyOnce`), while still having documentation in the manual for them

Can also do the same with all the mysql options that are mirrored under `cfg.database`

Reply:
I reorganized the options and migrated some of them to freeform modules. I kept some of the options unchanged for 1. settings with wide effects on the system (e.g., ports, database creation), 2. settings that are prevalent in other modules (e.g., ports), 3. settings that shouldn't appear at all in the resulting configuration file under some condition (e.g., basicAuth).

I also un-exposed some database options which most users don't need to change.